clients/drcachesim/scheduler/scheduler.h

```diff
@@ -556,7 +568,9 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
          * before knowing whether a simulator is on the wrong path or not, this routine
          * supports putting back the current record so that it will be re-provided as
          * the first record after (the outermost) stop_speculation(), if
-         * "queue_current_record" is true.
+         * "queue_current_record" is true.  An alternative way to accomplish this
```

Comment:
My reading of the comment suggests that start_speculation w/ queue_current_record=true supports arbitrary nesting, while unread_last_record() does not.  If so, that seems like a strong argument against using unread_last_record() as an alternative here.  Unless start_speculation() will _also_ nest multiple unread last records.
For that matter, it's a little unclear what the interaction is between unread_last_record() and start_speculation().  What happens if you call unread_last_record(), start_speculation(), and then next_record()?  Do you get the unread record or the side stream?

Reply:
On further examination, you are right: the interactions do get messy.  I think the best thing is to disallowed mixing them: have unread_last_record() fail during speculation and have start_speculation(queue_current_record=true) fail if unread_last_record() was just called.  I documented this and added sanity checks for each failure case.